File: cpp/okf/parse.cpp
// Frontmatter parsing + body normalization — mirrors py/okf/okf.py::parse_file.
//
// Parity notes:
// - Lines split like Python splitlines() / R readLines(): EOL stripped
//   (including a trailing newline, and the CR of CRLF), so the body — and
//   therefore content_hash — is byte-identical across bindings.
// - YAML via rapidyaml: scalars are raw text (no implicit typing at parse
//   time), so `timestamp: 2026-06-22T00:00:00Z` stays a verbatim string with
//   no custom-loader work, and `okf_version: 0.1` keeps its "0.1" text.
//   Frontmatter scalars therefore land in JSON as strings (semantically
//   equal, not byte-locked — see docs/ARCHITECTURE.md).

#include <mutex>

#include <ryml.hpp>
#include <ryml_std.hpp>

#include "okf.hpp"
#include "sha1.hpp"

namespace okf {
namespace {

std::vector<std::string> split_lines(const std::string& text) {
    std::vector<std::string> lines;
    std::size_t start = 0;
    while (start <= text.size()) {
        std::size_t nl = text.find('\n', start);
        if (nl == std::string::npos) {
            if (start < text.size()) lines.push_back(text.substr(start));
            break;
        }
        std::string line = text.substr(start, nl - start);
        if (!line.empty() && line.back() == '\r') line.pop_back();
        lines.push_back(std::move(line));
        start = nl + 1;
    }
    return lines;
}

bool is_fence(const std::string& line) {  // ^---\s*$
    if (line.rfind("---", 0) != 0) return false;
    return line.find_first_not_of(" \t\f\v\r", 3) == std::string::npos;
}

bool is_blank(const std::string& line) {
    return line.find_first_not_of(" \t\f\v\r") == std::string::npos;
}

std::string join(const std::vector<std::string>& lines, std::size_t from, std::size_t to) {
    std::string out;
    for (std::size_t i = from; i < to; ++i) {
        if (i > from) out.push_back('\n');
        out += lines[i];
    }
    return out;
}

// rapidyaml aborts on parse errors by default; route errors into exceptions
// once, process-wide.
void install_ryml_error_handler() {
    static std::once_flag flag;
    std::call_once(flag, [] {
        ryml::Callbacks cb = ryml::get_callbacks();
        cb.m_error = [](const char* msg, std::size_t len, ryml::Location, void*) {
            throw std::runtime_error(std::string(msg, len));
        };
        ryml::set_callbacks(cb);
    });
}

json yaml_to_json(ryml::ConstNodeRef n) {
    if (n.is_map()) {
        json obj = json::object();
        for (ryml::ConstNodeRef ch : n.children()) {
            std::string key(ch.key().str, ch.key().len);
            obj[key] = yaml_to_json(ch);
        }
        return obj;
    }
    if (n.is_seq()) {
        json arr = json::array();
        for (ryml::ConstNodeRef ch : n.children()) arr.push_back(yaml_to_json(ch));
        return arr;
    }
    if (!n.has_val() || n.val_is_null()) return nullptr;
    return std::string(n.val().str, n.val().len);
}

}  // namespace

Parsed parse_text(const std::string& text) {
    install_ryml_error_handler();
    std::vector<std::string> raw = split_lines(text);
    std::string txt = join(raw, 0, raw.size());
    std::size_t i = 0;
    while (i < raw.size() && is_blank(raw[i])) ++i;
    if (i >= raw.size() || !is_fence(raw[i])) {
        return Parsed{nullptr, txt, std::string("no_frontmatter")};
    }
    std::size_t opn = i, close = raw.size();
    for (std::size_t j = opn + 1; j < raw.size(); ++j) {
        if (is_fence(raw[j])) { close = j; break; }
    }
    if (close == raw.size()) {
        return Parsed{nullptr, txt, std::string("unclosed_frontmatter")};
    }
    std::string fm = join(raw, opn + 1, close);
    std::string body = close + 1 < raw.size() ? join(raw, close + 1, raw.size()) : "";
    json meta = nullptr;
    try {
        ryml::Tree tree = ryml::parse_in_arena(ryml::to_csubstr(fm));
        ryml::ConstNodeRef root = tree.rootref();
        if (root.is_map()) meta = yaml_to_json(root);
    } catch (const std::exception&) {
        meta = nullptr;
    }
    if (!meta.is_object()) {
        return Parsed{nullptr, std::move(body), std::string("yaml_parse_error")};
    }
    return Parsed{std::move(meta), std::move(body), std::nullopt};
}

std::string content_hash(const std::string& body) { return detail::sha1_hex(body); }

}  // namespace okf

File: rust/okf-ingest/src/parse.rs
//! Frontmatter parsing + body normalization — mirrors `py/okf/okf.py::parse_file`.
//!
//! Parity notes:
//! - Lines are split like Python `splitlines()` / R `readLines()`: the EOL is
//!   stripped (including a trailing newline, and the CR of CRLF), so the body —
//!   and therefore `content_hash` — is byte-identical across bindings.
//! - YAML is loaded with yaml-rust2 (YAML 1.2 core schema): there is no
//!   timestamp tag, so `timestamp: 2026-06-22T00:00:00Z` stays a verbatim
//!   string with no custom-loader work (the Python binding strips the resolver
//!   explicitly; see `_OKFLoader`).

use sha1::{Digest, Sha1};
use yaml_rust2::yaml::Hash;
use yaml_rust2::{Yaml, YamlLoader};

/// Result of parsing one document's text.
#[derive(Debug)]
pub struct Parsed {
    pub meta: Option<Hash>,
    pub body: String,
    /// `no_frontmatter` | `unclosed_frontmatter` | `yaml_parse_error`.
    pub err: Option<&'static str>,
}

fn is_fence(line: &str) -> bool {
    // ^---\s*$
    line.strip_prefix("---")
        .is_some_and(|rest| rest.trim().is_empty())
}

/// Parse frontmatter + body from raw file text.
pub fn parse_text(text: &str) -> Parsed {
    // str::lines() == Python splitlines() for \n and \r\n (strips trailing
    // newline; no phantom empty final line).
    let raw: Vec<&str> = text.lines().collect();
    let txt = raw.join("\n");
    let mut i = 0;
    while i < raw.len() && raw[i].trim().is_empty() {
        i += 1;
    }
    if i >= raw.len() || !is_fence(raw[i]) {
        return Parsed {
            meta: None,
            body: txt,
            err: Some("no_frontmatter"),
        };
    }
    let opn = i;
    let close = match (opn + 1..raw.len()).find(|&j| is_fence(raw[j])) {
        Some(c) => c,
        None => {
            return Parsed {
                meta: None,
                body: txt,
                err: Some("unclosed_frontmatter"),
            }
        }
    };
    let fm = raw[opn + 1..close].join("\n");
    let body = if close < raw.len() - 1 {
        raw[close + 1..].join("\n")
    } else {
        String::new()
    };
    let meta = match YamlLoader::load_from_str(&fm) {
        Ok(docs) => docs.into_iter().next(),
        Err(_) => None,
    };
    match meta {
        Some(Yaml::Hash(h)) => Parsed {
            meta: Some(h),
            body,
            err: None,
        },
        _ => Parsed {
            meta: None,
            body,
            err: Some("yaml_parse_error"),
        },
    }
}

/// sha1 hex of the normalized body (the cross-language parity lock).
pub fn content_hash(body: &str) -> String {
    let digest = Sha1::digest(body.as_bytes());
    digest.iter().fold(String::with_capacity(40), |mut s, b| {
        use std::fmt::Write;
        let _ = write!(s, "{b:02x}");
        s
    })
}

/// Scalar-to-string coercion — the `_s()` of the Python binding: `None` for
/// sequences/mappings/null, otherwise the scalar as a string. yaml-rust2 keeps
/// the raw text of floats (`0.1` stays `"0.1"`), matching Python `str(0.1)`.
pub fn scalar_str(y: &Yaml) -> Option<String> {
    match y {
        Yaml::String(s) => Some(s.clone()),
        Yaml::Real(s) => Some(s.clone()),
        Yaml::Integer(i) => Some(i.to_string()),
        Yaml::Boolean(b) => Some(b.to_string()),
        _ => None,
    }
}

/// YAML value -> JSON value (for `tags` and the `frontmatter` column). The
/// JSON serialization of `tags` is what lexical seeding matches against.
pub fn yaml_to_json(y: &Yaml) -> serde_json::Value {
    use serde_json::Value;
    match y {
        Yaml::Null | Yaml::BadValue | Yaml::Alias(_) => Value::Null,
        Yaml::Boolean(b) => Value::Bool(*b),
        Yaml::Integer(i) => Value::from(*i),
        Yaml::Real(s) => match s.parse::<f64>() {
            Ok(f) => serde_json::Number::from_f64(f)
                .map(Value::Number)
                .unwrap_or(Value::Null),
            Err(_) => Value::String(s.clone()),
        },
        Yaml::String(s) => Value::String(s.clone()),
        Yaml::Array(a) => Value::Array(a.iter().map(yaml_to_json).collect()),
        Yaml::Hash(h) => {
            let mut m = serde_json::Map::new();
            for (k, v) in h.iter() {
                let key = scalar_str(k).unwrap_or_default();
                m.insert(key, yaml_to_json(v));
            }
            Value::Object(m)
        }
    }
}

/// Fetch a frontmatter value by string key.
pub fn meta_get<'a>(h: &'a Hash, key: &str) -> Option<&'a Yaml> {
    h.get(&Yaml::String(key.to_string()))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn fence_matching() {
        assert!(is_fence("---"));
        assert!(is_fence("---  "));
        assert!(!is_fence("----"));
        assert!(!is_fence(" ---"));
        assert!(!is_fence("--- x"));
    }

    #[test]
    fn crlf_and_trailing_newline_normalize() {
        let unix = "---\ntype: Signal\n---\nbody line\n";
        let dos = "---\r\ntype: Signal\r\n---\r\nbody line\r\n";
        let p1 = parse_text(unix);
        let p2 = parse_text(dos);
        assert_eq!(p1.body, "body line");
        assert_eq!(p1.body, p2.body);
        assert_eq!(content_hash(&p1.body), content_hash(&p2.body));
    }

    #[test]
    fn no_frontmatter_and_unclosed() {
        assert_eq!(parse_text("just text").err, Some("no_frontmatter"));
        assert_eq!(parse_text("---\ntype: X").err, Some("unclosed_frontmatter"));
        assert_eq!(
            parse_text("---\n- a\n- b\n---\nx").err,
            Some("yaml_parse_error")
        );
    }

    #[test]
    fn timestamp_stays_verbatim() {
        let p = parse_text("---\ntimestamp: 2026-06-22T00:00:00Z\n---\nx");
        let m = p.meta.unwrap();
        let v = meta_get(&m, "timestamp").unwrap();
        assert_eq!(scalar_str(v).as_deref(), Some("2026-06-22T00:00:00Z"));
    }

    #[test]
    fn float_scalar_keeps_raw_text() {
        let p = parse_text("---\nokf_version: 0.1\n---\nx");
        let m = p.meta.unwrap();
        assert_eq!(
            scalar_str(meta_get(&m, "okf_version").unwrap()).as_deref(),
            Some("0.1")
        );
    }
}
